File: ui/src/mocks/handlers/slos.ts
// src/mocks/handlers/slos.ts
import { http, HttpResponse } from 'msw'

async function gen() {
  return import('../generate')
}

export const sloHandlers = [
  http.post('/api/slo-definitions/validate', async ({ request }) => {
    await request.json()
    return HttpResponse.json({
      valid: true,
      errors: [],
      objectives: [
        { sli: 'response_time_p95', display_name: 'Response Time P95', pass_threshold: ['<500'], warning_threshold: ['<800'], weight: 1, key_sli: false, sort_order: 0 },
        { sli: 'error_rate', display_name: 'Error Rate', pass_threshold: ['<=0.5%'], warning_threshold: ['<=2%'], weight: 1, key_sli: false, sort_order: 1 },
      ],
    })
  }),

  http.post('/api/slo-definitions/test', async () => {
    return HttpResponse.json({
      result: 'pass',
      score: 91.5,
      indicator_results: [],
      baseline_mode: 'none',
      metrics_fetched: {},
      fetch_errors: {},
      compared_values: null,
    })
  }),

  http.get('/api/slo-definitions', async () => {
    const { getSloDefinitions } = await gen()
    const items = getSloDefinitions()
    return HttpResponse.json({ items, total: (items as unknown[]).length })
  }),

  http.get('/api/slo-definitions/:name/versions', async ({ params }) => {
    const { getSloDefinitions } = await gen()
    const all = getSloDefinitions() as { name: string; version: number; created_at: string; author?: string | null; notes?: string | null; active: boolean }[]
    const current = all.find(s => s.name === params.name)
    if (!current) return HttpResponse.json({ error: 'Not found' }, { status: 404 })
    const history = [
      { ...current },
      {
        ...current,
        version: current.version - 1,
        active: false,
        notes: current.version > 1 ? `Previous v${current.version - 1}` : null,
        created_at: new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString(),
      },
    ].filter(v => v.version > 0)
    return HttpResponse.json(history)
  }),

  http.get('/api/slo-definitions/:name', async ({ params }) => {
    const { getSloDefinitions } = await gen()
    const all = getSloDefinitions()
    const slo = (all as { name: string }[]).find(s => s.name === params.name)
    if (!slo) return HttpResponse.json({ error: 'Not found' }, { status: 404 })
    return HttpResponse.json(slo)
  }),

  http.delete('/api/slo-definitions/:name', async ({ params }) => {
    console.log('[mock] soft-delete SLO:', params.name)
    return new HttpResponse(null, { status: 204 })
  }),

  http.post('/api/slo-definitions', async ({ request }) => {
    const body = await request.json() as {
      name: string
      objectives: unknown[]
      total_score_pass_pct: number
      total_score_warning_pct: number
      comparison: Record<string, unknown>
      display_name?: string
      notes?: string
      author?: string
    }
    return HttpResponse.json({
      id: crypto.randomUUID(),
      name: body.name,
      version: 1,
      display_name: body.display_name ?? null,
      author: body.author ?? null,
      notes: body.notes ?? null,
      active: true,
      tags: {},
      variables: {},
      created_at: new Date().toISOString(),
      objectives: body.objectives,
      total_score_pass_pct: body.total_score_pass_pct,
      total_score_warning_pct: body.total_score_warning_pct,
      comparison: body.comparison,
    }, { status: 201 })
  }),
]

File: clients/python/tropek_client/models/slos.py
"""SLO definition models."""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING, Any
from uuid import UUID

from pydantic import BaseModel, Field

if TYPE_CHECKING:
    from tropek_client.models.evaluations import IndicatorResult

from tropek_client.models.common import AggregateFunction


class ChangePointConfigInput(BaseModel):
    """Change point detection config for SLO objective (create/update)."""

    enabled: bool | None = None
    higher_is_better: bool | None = None
    window_size: int | None = None
    max_pvalue: float | int | None = None
    min_magnitude: float | int | None = None
    min_sample_size: int | None = None


class ChangePointConfigRead(BaseModel):
    """Change point detection config as returned by the API."""

    slo_objective_id: UUID
    enabled: bool
    higher_is_better: bool
    window_size: int
    max_pvalue: float | int
    min_magnitude: float | int
    min_sample_size: int


class SLOObjectiveIn(BaseModel):
    """SLO objective for create/update requests."""

    sli: str
    display_name: str = ''
    pass_threshold: list[str] | None = None
    warning_threshold: list[str] | None = None
    weight: int = 1
    key_sli: bool = False
    change_point: ChangePointConfigInput | None = None


class SLOObjectiveRead(BaseModel):
    """SLO objective as returned by the API."""

    sli: str
    display_name: str = ''
    pass_threshold: list[str] | None = None
    warning_threshold: list[str] | None = None
    weight: int = 1
    key_sli: bool = False
    sort_order: int
    change_point: ChangePointConfigRead | None = None


class ComparisonConfig(BaseModel):
    """Comparison configuration for SLO create/update."""

    compare_with: str | None = None
    include_result_with_score: str | None = None
    number_of_comparison_results: int | None = None
    aggregate_function: AggregateFunction | None = None
    scope_tags: list[str] | None = None


class ComparisonConfigRead(BaseModel):
    """Comparison configuration as returned by the API."""

    compare_with: str | None = None
    include_result_with_score: str | None = None
    number_of_comparison_results: int | None = None
    aggregate_function: str | None = None
    scope_tags: list[str] | None = None


class MethodCriteriaOverride(BaseModel):
    """Per-method criteria override for SLO create/update."""

    method: str | None = None
    aggregation: str | None = None
    pass_threshold: list[str] | None = None
    warning_threshold: list[str] | None = None
    weight: int | None = None
    key_sli: bool | None = None


class MethodCriteriaOverrideRead(BaseModel):
    """Per-method criteria override as returned by the API."""

    method: str | None = None
    aggregation: str | None = None
    pass_threshold: list[str] | None = None
    warning_threshold: list[str] | None = None
    weight: int | None = None
    key_sli: bool | None = None


class BaselineConfig(BaseModel):
    """Baseline configuration for SLO test requests."""

    mode: str = 'none'
    limit: int = 3
    values: dict[str, Any] | None = None


CompareToValue = bool | str


class ComparisonRule(BaseModel):
    """Rule mapping tag match criteria to comparison target values."""

    match: dict[str, str]
    compare_to: dict[str, CompareToValue]


class SLODefinitionCreate(BaseModel):
    """Request body for creating a new SLO definition."""

    name: str
    display_name: str | None = None
    objectives: list[SLOObjectiveIn]
    total_score_pass_threshold: float | int = 90.0
    total_score_warning_threshold: float | int = 75.0
    comparison: ComparisonConfig | None = None
    notes: str | None = None
    author: str | None = None
    tags: dict[str, str] | None = None
    variables: dict[str, str] | None = None
    comparable_from_version: int | None = None
    kind: str = 'standard'
    sli_name: str | None = None
    sli_version: int | None = None
    method_criteria: dict[str, MethodCriteriaOverride] | None = None


class SLODefinitionRead(BaseModel):
    """SLO definition as returned by the API."""

    id: UUID
    name: str
    display_name: str | None = None
    version: int
    comparable_from_version: int
    active: bool
    objectives: list[SLOObjectiveRead]
    total_score_pass_threshold: float | int
    total_score_warning_threshold: float | int
    comparison: ComparisonConfigRead
    notes: str | None = None
    author: str | None = None
    tags: dict[str, str]
    variables: dict[str, str]
    kind: str
    method_criteria: dict[str, MethodCriteriaOverrideRead] | None = None
    sli_definition_id: UUID | None = None
    sli_name: str | None = None
    sli_version: int | None = None
    created_at: datetime


class SLOValidateRequest(BaseModel):
    """Request body for SLO validation."""

    objectives: list[SLOObjectiveIn]
    total_score_pass_threshold: float | int = 90.0
    total_score_warning_threshold: float | int = 75.0
    comparison: ComparisonConfig | None = None


class SLOValidationError(BaseModel):
    """Single validation error returned by the SLO validate endpoint."""

    var_field: str = Field(alias='field')
    message: str


class SLOValidationResult(BaseModel):
    """Response from the SLO validate endpoint."""

    valid: bool
    errors: list[SLOValidationError]
    objectives: list[SLOObjectiveIn] | None = None


class SLOTestRequest(BaseModel):
    """Request body for testing an SLO against live data."""

    objectives: list[SLOObjectiveIn]
    total_score_pass_threshold: float | int = 90.0
    total_score_warning_threshold: float | int = 75.0
    comparison: ComparisonConfig | None = None
    sli_name: str
    data_source_name: str
    asset_name: str
    period_start: datetime
    period_end: datetime
    evaluation_name: str = ''
    baseline: BaselineConfig | None = None
    variables: dict[str, str] | None = None


class SLOTestResult(BaseModel):
    """Response from the SLO test endpoint."""

    result: str
    score: float | int
    indicator_results: list[IndicatorResult]
    baseline_mode: str
    metrics_fetched: dict[str, float | int]
    fetch_errors: dict[str, str]
    compared_values: dict[str, float | int] | None = None
